### PR TITLE
Adding code for the ARM LRO state machine

### PR DESCRIPTION
**Pull request by catalinaperalta**
- [ ] The purpose of this PR is explained in this or a referenced issue.
- [ ] The PR does not update generated files.
   - These files are managed by the codegen framework at [Azure/autorest.go][].
- [ ] Tests are included and/or updated for code changes.
- [ ] Updates to [CHANGELOG.md][] are included.
- [ ] Apache v2 license headers are included in each file.
 
[Azure/autorest.go]: https://github.com/Azure/autorest.go
[CHANGELOG.md]: https://github.com/Azure/azure-sdk-for-go/blob/master/CHANGELOG.md

### CODE REVIEW

File: sdk/armcore/poller_test.go [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+		azcore.NewUniqueRequestIDPolicy(),
+		azcore.NewRetryPolicy(nil),
+		azcore.NewRequestLogPolicy(azcore.RequestLogOptions{}))
+}
+
+func handleError(resp *azcore.Response) error {
+	return fmt.Errorf("error status: %d", resp.StatusCode)
+}
+
+func TestCreatePollerTracker(t *testing.T) {
```
Comment (jhendrixMSFT) [inline sdk/armcore/poller_test.go:40]: The great thing about this is we now have code coverage info for the state machine.  Looks like we need a lot more tests (can do this in a separate PR).

File: sdk/armcore/poller.go [hunk 2]
```diff
@@ -0,0 +1,1068 @@
...
+	operationFailed     string = "Failed"
+	operationSucceeded  string = "Succeeded"
+)
+
+var pollingCodes = [...]int{http.StatusNoContent, http.StatusAccepted, http.StatusCreated, http.StatusOK}
+
+// CreatePollingTracker creates a polling tracker based on the verb of the original request and returns
+// the polling tracker implementation for the method verb or an error.
+// NOTE: this is only meant for internal use in generated code.
+func CreatePollingTracker(pollerType string, finalState string, resp *azcore.Response, errorHandler methodErrorHandler) (Poller, error) {
```
Comment (jhendrixMSFT) [inline sdk/armcore/poller.go:39]: I would rename this to `CreatePoller` as it returns a `Poller`.  The whole tracker thing is really an implementation detail.

File: sdk/armcore/poller.go [hunk 3]
```diff
@@ -0,0 +1,1068 @@
...
+		// error (not doing this means the error shows up in Future.Done)
+		return pt, pt.updatePollingMethod()
+	default:
+		return nil, fmt.Errorf("unsupported HTTP method %s", resp.Request.Method)
+	}
+}
+
+// ResumePollingTracker creates a polling tracker from a resume token string.
+// NOTE: this is only meant for internal use in generated code.
+func ResumePollingTracker(pollerType string, token string, errorHandler methodErrorHandler) (Poller, error) {
```
Comment (jhendrixMSFT) [inline sdk/armcore/poller.go:84]: Same here `ResumePoller`

File: sdk/armcore/poller.go [hunk 4]
```diff
@@ -0,0 +1,1068 @@
...
+}
+
+// Poller defines the methods that will be called internally in the generated code for long-running operations.
+// NOTE: this is only meant for internal use in generated code.
+type Poller interface {
+	// Done signals if the polling operation has reached a terminal state.
+	Done() bool
+	// Poll sends a polling request to the service endpoint and returns the http.Response received from the endpoint or an error.
+	Poll(ctx context.Context, p azcore.Pipeline) (*http.Response, error)
+	// FinalResponse will perform a final GET and return the final response for the polling operation.
```
Comment (jhendrixMSFT) [inline sdk/armcore/poller.go:143]: Comment needs slight rewording to accommodate the `respType` param.

File: sdk/armcore/poller.go [hunk 5]
```diff
@@ -0,0 +1,1068 @@
...
+	FinalResponse(ctx context.Context, pipeline azcore.Pipeline, respType interface{}) (*http.Response, error)
+	// ResumeToken returns a token string that can be used to resume polling on a poller that has not yet reached a terminal state.
+	ResumeToken() (string, error)
+	// PollUntilDone will handle the entire span of the polling operation until a terminal state is reached.
```
Comment (jhendrixMSFT) [inline sdk/armcore/poller.go:147]: Similar rewording as above.

File: sdk/armcore/poller_test.go (hunk 1 above)
Comment (catalinaperalta) [reply sdk/armcore/poller_test.go:40]: Yeah agreed, I was thinking about how far to go with code coverage here vs in the generator which manages to cover many different code paths.

File: sdk/armcore/poller.go (hunk 2 above)
Comment (catalinaperalta) [reply sdk/armcore/poller.go]: Done

File: sdk/armcore/poller.go (hunk 3 above)
Comment (catalinaperalta) [reply sdk/armcore/poller.go]: Done

File: sdk/armcore/poller.go (hunk 4 above)
Comment (catalinaperalta) [reply sdk/armcore/poller.go]: Done

File: sdk/armcore/poller.go (hunk 5 above)
Comment (catalinaperalta) [reply sdk/armcore/poller.go]: Done

File: sdk/armcore/poller.go [hunk 6]
```diff
@@ -0,0 +1,1072 @@
...
+// NOTE: this is only meant for internal use in generated code.
+func ResumePoller(pollerType string, token string, errorHandler methodErrorHandler) (Poller, error) {
+	// unmarshal into JSON object to determine the tracker type
+	obj := map[string]interface{}{}
+	err := json.Unmarshal([]byte(token), &obj)
+	if err != nil {
+		return nil, err
+	}
+	if obj["pollerType"] != pollerType {
+		return nil, fmt.Errorf("cannot resume from this poller type")
```
Comment (jhendrixMSFT) [inline sdk/armcore/poller.go:92]: Probably good to include the value of `obj["pollerType"]` for extra diagnostics.

File: sdk/armcore/poller.go [hunk 7]
```diff
@@ -0,0 +1,1072 @@
...
+func (p *pollingTrackerBase) FinalResponse(ctx context.Context, pipeline azcore.Pipeline, respType interface{}) (*http.Response, error) {
+	if !p.Done() {
+		return nil, errors.New("cannot return a final response from a poller in a non-terminal state")
+	}
+	if p.pollerMethodVerb() == http.MethodPut || p.pollerMethodVerb() == http.MethodPatch {
+		res, err := p.handleResponse(p.latestResponse(), respType)
+		if err != nil {
+			return nil, err
+		}
+		if res != nil { // TODO && res.ProductArray != nil {
```
Comment (jhendrixMSFT) [inline sdk/armcore/poller.go:166]: Is the trailing comment here important?

Comment (catalinaperalta) [reply sdk/armcore/poller.go:166]: Yes it it :( looking into it. The implementation for here might need to change

Comment (catalinaperalta) [reply sdk/armcore/poller.go]: Done

File: sdk/armcore/poller.go (hunk 6 above)
Comment (catalinaperalta) [reply sdk/armcore/poller.go]: Done

**Comment by catalinaperalta** [comment]
More tests to increase code coverage to come in the next PR

File: sdk/armcore/poller.go [hunk 8]
```diff
@@ -0,0 +1,1073 @@
...
+	operationFailed     string = "Failed"
+	operationSucceeded  string = "Succeeded"
+)
+
+var pollingCodes = [...]int{http.StatusNoContent, http.StatusAccepted, http.StatusCreated, http.StatusOK}
+
+// CreatePoller creates a polling tracker based on the verb of the original request and returns
+// the polling tracker implementation for the method verb or an error.
+// NOTE: this is only meant for internal use in generated code.
+func CreatePoller(pollerType string, finalState string, resp *azcore.Response, errorHandler methodErrorHandler) (Poller, error) {
```
Comment (jhendrixMSFT) [inline sdk/armcore/poller.go:40]: More thinking, this should be `NewPoller()` like our other ctors.

File: sdk/armcore/poller.go [hunk 9]
```diff
@@ -0,0 +1,1073 @@
...
+		// error (not doing this means the error shows up in Future.Done)
+		return pt, pt.updatePollingMethod()
+	default:
+		return nil, fmt.Errorf("unsupported HTTP method %s", resp.Request.Method)
+	}
+}
+
+// ResumePoller creates a polling tracker from a resume token string.
+// NOTE: this is only meant for internal use in generated code.
+func ResumePoller(pollerType string, token string, errorHandler methodErrorHandler) (Poller, error) {
```
Comment (jhendrixMSFT) [inline sdk/armcore/poller.go:85]: `NewPollerFromResumeToken()` ?

Comment (catalinaperalta) [reply sdk/armcore/poller.go]: Done

File: sdk/armcore/poller.go (hunk 8 above)
Comment (catalinaperalta) [reply sdk/armcore/poller.go]: Done

File: sdk/armcore/poller.go [hunk 10]
```diff
@@ -0,0 +1,1077 @@
...
+	return p.hasTerminated()
+}
+
+// FinalResponse will perform a final GET and return the final response for the polling operation.
+func (p *pollingTrackerBase) FinalResponse(ctx context.Context, pipeline azcore.Pipeline, respType interface{}) (*http.Response, error) {
+	if !p.Done() {
+		return nil, errors.New("cannot return a final response from a poller in a non-terminal state")
+	}
+	// if respType is nil, this indicates that the request was made from an HTTPPoller
+	if respType == nil {
```
Comment (catalinaperalta) [inline sdk/armcore/poller.go:163]: FYI this is new and necessary for HTTPPollers

Comment (jhendrixMSFT) [reply sdk/armcore/poller.go:163]: In that case can we simply bypass the call to FinalResponse()?

Comment (catalinaperalta) [reply sdk/armcore/poller.go:163]: No, since we dont expose anything else aside from the Poller interface methods

File: sdk/armcore/go.mod [hunk 11]
```diff
@@ -3,6 +3,6 @@ module github.com/Azure/azure-sdk-for-go/sdk/armcore
 go 1.13
 
 require (
-	github.com/Azure/azure-sdk-for-go/sdk/azcore v0.9.0
-	github.com/Azure/azure-sdk-for-go/sdk/internal v0.2.0
+	github.com/Azure/azure-sdk-for-go/sdk/azcore v0.9.3
```
Comment (jhendrixMSFT) [inline sdk/armcore/go.mod:6]: Could you please bump this to 0.9.4 (released this morning).

Comment (catalinaperalta) [reply sdk/armcore/go.mod]: Done
